File: QytetetRuby/lib/tablero.rb
#Elena Cantero Molina
#encoding: utf-8
#trig.rb

module ModeloQytetet
  class Tablero
    attr_reader :carcel, :casillas
    
    def initialize
      @casillas = Array.new
      @casillas << Casilla.new(0, 0, TipoCasilla::SALIDA)
      @casillas << Calle.new(1, 300, TituloPropiedad.new("Calle Tranvia", 50, 0.10, 250, 150))
      @casillas << Casilla.new(2, -200, TipoCasilla::IMPUESTO)
      @casillas << Calle.new(3, 400, TituloPropiedad.new("Calle Vengala", 55, -0.16, 280, 300))
      @casillas << Calle.new(4, 450, TituloPropiedad.new("Calle Tren", 60, 0.18, 300, 450))
      @casillas << Casilla.new(5, 0, TipoCasilla::CARCEL)
      @casillas << Calle.new(6, 500, TituloPropiedad.new("Calle Final", 60, 0.20, 350, 500))
      @casillas << Casilla.new(7, 0, TipoCasilla::SORPRESA)
      @casillas << Calle.new(8, 550, TituloPropiedad.new("Calle Reinado", 65, -0.15, 400, 600))
      @casillas << Calle.new(9, 650, TituloPropiedad.new("Calle Europa", 70, 0.17, 450, 650))
      @casillas << Casilla.new(10, 0, TipoCasilla::PARKING)
      @casillas << Calle.new(11, 700, TituloPropiedad.new("Calle Periodistas", 70, 0.14, 500, 700))
      @casillas << Calle.new(12, 700, TituloPropiedad.new("Calle Cabezas", 75, 0.16, 550, 750))
      @casillas << Casilla.new(13, 0, TipoCasilla::SORPRESA)
      @casillas << Calle.new(14, 750, TituloPropiedad.new("Calle Molina", 80, -0.20, 600, 800))
      @casillas << Casilla.new(15, 0, TipoCasilla::JUEZ)
      @casillas << Calle.new(16, 800, TituloPropiedad.new("Calle America", 90, 0.18, 650, 850))
      @casillas << Calle.new(17, 900, TituloPropiedad.new("Calle Humilladero", 95, 0.12, 700, 900))
      @casillas << Casilla.new(18, 0, TipoCasilla::SORPRESA)
      @casillas << Calle.new(19, 950, TituloPropiedad.new("Calle Princesa", 100, -0.19, 750, 950))
      @carcel = @casillas[5] 
    end
    
    def es_casilla_carcel(numero_casilla)
      return numero_casilla.eql?(@carcel.numero_casilla)
    end    

    #precondicion -> numeroCasilla > 0 && numeroCasilla < 20
    def obtener_casilla_numero(numero_casilla)
      @casillas.at(numero_casilla)
    end    

    def obtener_nueva_casilla(casilla, desplazamiento)
      numero_casilla = casilla.numero_casilla
      numero_casilla = (numero_casilla + desplazamiento) % @casillas.size
      return @casillas.at(numero_casilla)
    end
    
    def to_s
      "Tablero: \nCasillas:\n#{@casillas}\nCarcel= #{@carcel}"
    end
  end
end

File: QytetetJava/src/modeloqytetet/Tablero.java
/**
 * @author Elena Cantero Molina
 */
package modeloqytetet;
import java.util.*;

public class Tablero {
    private ArrayList<Casilla> casillas;    //Array para contener las casillas
    private Casilla carcel; //indica la cárcel
    
    public Tablero(){
        inicializar();
    }
    
    private void inicializar(){
        casillas = new ArrayList();
        
        casillas.add(new OtraCasilla(0, 0, TipoCasilla.SALIDA));
        casillas.add(new Calle(1, 300, new TituloPropiedad("Calle Tranvia", 50, 10, 250, 150)));
        casillas.add(new OtraCasilla(2, -100, TipoCasilla.IMPUESTO));
        casillas.add(new Calle(3, 400, new TituloPropiedad("Calle Vengala", 55, 16, 280, 300)));
        casillas.add(new Calle(4, 450, new TituloPropiedad("Calle Tren", 60, 18, 300, 450)));
        casillas.add(new OtraCasilla(5, 0, TipoCasilla.CARCEL));
        casillas.add(new Calle(6, 500, new TituloPropiedad("Calle Final", 60, 20, 350, 500)));
        casillas.add(new OtraCasilla(7, 0, TipoCasilla.SORPRESA));
        casillas.add(new Calle(8, 550, new TituloPropiedad("Calle Reinado", 65, 15, 400, 600)));
        casillas.add(new Calle(9, 650, new TituloPropiedad("Calle Europa", 70, 17, 450, 650)));
        casillas.add(new OtraCasilla(10, 0, TipoCasilla.PARKING));
        casillas.add(new Calle(11, 700, new TituloPropiedad("Calle Periodistas", 70, 14, 500, 700)));
        casillas.add(new Calle(12, 700, new TituloPropiedad("Calle Cabezas", 75, 16, 550, 750)));
        casillas.add(new OtraCasilla(13, 0, TipoCasilla.SORPRESA));
        casillas.add(new Calle(14, 750, new TituloPropiedad("Calle Molina", 80, 20, 600, 800)));
        casillas.add(new OtraCasilla(15, 0, TipoCasilla.JUEZ));
        casillas.add(new Calle(16, 800, new TituloPropiedad("Calle America", 90, 18, 650, 850)));
        casillas.add(new Calle(17, 900, new TituloPropiedad("Calle Humilladero", 95, 12, 700, 900)));
        casillas.add(new OtraCasilla(18, 0, TipoCasilla.SORPRESA));
        casillas.add(new Calle(19, 950, new TituloPropiedad("Calle Princesa", 100, 19, 750, 950)));
        
        carcel = casillas.get(5);
    }
    
    Casilla GetCarcel(){ return carcel; }
    
    boolean esCasillaCarcel(int numeroCasilla){
        return numeroCasilla == this.GetCarcel().GetNumeroCasilla();
    }    

    //precondicion -> numeroCasilla > 0 && numeroCasilla < 20
    Casilla obtenerCasillaNumero(int numeroCasilla){
        return this.casillas.get(numeroCasilla);
    }    

    Casilla obtenerNuevaCasilla(Casilla casilla, int desplazamiento){
        int numero_casilla = casilla.GetNumeroCasilla();
        numero_casilla = (numero_casilla + desplazamiento) % casillas.size();
        return this.casillas.get(numero_casilla);
    }
    
    @Override
    public String toString(){
        return "Tablero -> \n" + this.casillas + "\nCarcel -> " + this.GetCarcel() + "\n";
    }
}
